File: game.cpp
//
// Created by cothrax on 1/5/21.
//

#include "calculator.h"
#include "game.h"
#include <algorithm>
#include <cmath>
#include <random>

InfoSet::InfoSet(Game &game)
{
    step = 0;
    memcpy(chips, game.chips, sizeof(chips));
    memset(folds, 0, sizeof(folds));
    memset(bets, 0, sizeof(bets));
    pot = SMALL_BLIND_CHIP + BIG_BLIND_CHIP;

    bets[game.start] = SMALL_BLIND_CHIP;
    bets[game.big_blind] = BIG_BLIND_CHIP;
    chips[game.start] -= SMALL_BLIND_CHIP;
    chips[game.big_blind] -= BIG_BLIND_CHIP;

    player = (game.start+2) % NUM_PLAYER;
    num = NUM_PLAYER;
}

void InfoSet::bet(int more){
    more = std::min(more, chips[player]);
    chips[player] -= more;
    bets[player] += more;
}

void Game::generate(int start_, int pov_)
{
    std::shuffle(cards, cards+52, std::mt19937(std::random_device()()));
    memcpy(pubs, cards, sizeof(pubs));
    memcpy(holes, cards + 5, sizeof(holes));

    pov = pov_;
    start = start_;
    big_blind = (start_ + 1) % NUM_PLAYER;
    std::fill(chips, chips+NUM_PLAYER, INIT_CHIP);

    for(int i = 0; i < NUM_PLAYER; i++)
        power[i] = calculator.power(holes[i], pubs);

    printf("gen game: start = %d, pov = %d\n\tholes: ", start_, pov_);
    for(int i = 0; i < NUM_PLAYER; i++) printf("(%d, %d), ", holes[i][0], holes[i][1]);
    printf("\n\tpubs:(%d, %d, %d)\n", pubs[0], pubs[1], pubs[2]);
}

int Game::change_state(InfoSet &info)
{
    if(info.num == 1) return GAME_OVER;
    int i, lst_bet = -1;
    for(i = 0; i < NUM_PLAYER; i++)
        if(!info.folds[i])
        {
            if(lst_bet == -1 || lst_bet == info.bets[i]) lst_bet = info.bets[i];
            else return NO_CHANGE;
        }

    info.player = big_blind;
    for(int bet : info.bets) info.pot += bet;
    memset(info.bets, 0, sizeof(info.bets));
    return ++info.step;
}

void Game::calc_result(InfoSet &info, double *util)
{
    int maxv = -1, cnt = 0;
    for(int i = 0; i < NUM_PLAYER; i++)
        if(!info.folds[i]) maxv = std::max(maxv, power[i]);
    for(int i = 0; i < NUM_PLAYER; i++)
        cnt += !info.folds[i] && power[i] == maxv;

    for(int i = 0; i < NUM_PLAYER; i++)
        if(!info.folds[i] && power[i] == maxv) util[i] = info.pot / cnt;
        else util[i] = chips[i] - info.chips[i];
}

void Game::act(InfoSet &info, int action, InfoSet &next_info)
{
    memcpy(&next_info, &info, sizeof(InfoSet));
    int prev_player = info.player, next_player = info.player;

    do prev_player = prev_player ? prev_player - 1 : NUM_PLAYER - 1;
    while(info.folds[prev_player]);
    do next_player = next_player == NUM_PLAYER - 1 ? 0 : next_player + 1;
    while(info.folds[next_player]);

    int det = info.bets[prev_player] - info.bets[info.player];
    int pot = next_info.pot;
    for(int i = 0; i < NUM_PLAYER; i++) pot += info.bets[i];

    switch(action) {
        case FOLD: next_info.folds[info.player] = true; next_info.num--; break;
        case CHECK: next_info.bet(det); break;
        case RAISE_3BB: next_info.bet(det + 3 * BIG_BLIND_CHIP); break;
        case RAISE_HALF_POT: next_info.bet(det + pot / 2); break;
        case RAISE_POT: next_info.bet(det + pot); break;
        case RAISE_2POT: next_info.bet(det + pot * 2); break;
        default: next_info.bet(INF32); break;
    }

    next_info.player = next_player;
}

File: cfr-py/game.py

import numpy as np
import random


NUM_PLAYER = 6
NUM_ACTION = 7
INIT_CHIPS = 20000
SMALL_BLIND = 50
BIG_BLIND = 100

FOLD = 0
CHECK = 1
RAISE_3BB = 6
RAISE_HALF_POT = 2
RAISE_POT = 4
RAISE_2POT = 3
ALL_IN = 5
NUM_ACTION = 7

NO_CHANGE = 0
PRE_FLOP = 0
FLOP = 1
TURN = 2
RIVER = 3
GAME_OVER = 4
NUM_STATE = 5

calculator = Calculator()


class Game:
    __slots__ = ['holes', 'pubs', 'start', 'big_blind', 'bets',
                 'chips', 'folds', 'power', 'step', 'num', 'pot',
                 'cur_bet', 'player']

    def __init__(self, start):
        self.holes = -np.ones(shape=(NUM_PLAYER, 2), dtype=np.int)
        self.pubs = -np.ones(shape=5, dtype=np.int)
        self.start = start
        self.big_blind = (start + 1) % NUM_PLAYER

        self.bets = np.zeros(shape=NUM_PLAYER, dtype=np.int)
        self.chips = np.ones(shape=NUM_PLAYER, dtype=np.int) * INIT_CHIPS
        self.folds = np.zeros(shape=NUM_PLAYER, dtype=np.bool)
        self.power = np.zeros(shape=NUM_PLAYER, dtype=np.bool)

        self.step = 0
        self.num = 5
        self.pot = 0
        self.cur_bet = BIG_BLIND

        self.bets[self.start] = SMALL_BLIND
        self.chips[self.start] -= SMALL_BLIND
        self.bets[self.big_blind] = BIG_BLIND
        self.chips[self.start] -= BIG_BLIND

        self.player = (self.big_blind + 1) % NUM_PLAYER
        self.generate()

    def generate(self):
        cards = np.random.choice(range(52), NUM_PLAYER*2+5, False)
        self.holes = cards[NUM_PLAYER*2].reshape(NUM_PLAYER, 2)
        self.pubs = cards[-5:]

        for i in range(NUM_PLAYER):
            self.power[i] = calculator.power(*self.holes[NUM_PLAYER], *self.pubs)

    def bet(self, more):
        more = min(more, self.chips[self.player])
        self.chips[self.player] -= more
        self.bets[self.player] += more

    def act(self, a):
        det = self.cur_bet - self.bets[self.player]
        pot = self.pot + np.sum(self.bets)
        if a == FOLD:
            self.folds[self.player] = True
            self.num -= 1
        elif a == CHECK:
            self.bet(det)
        elif a == RAISE_3BB:
            self.bet(det + BIG_BLIND * 3)
        elif a == RAISE_POT:
            self.bet(det + pot)
        elif a == RAISE_HALF_POT:
            self.bet(det + pot // 2)
        elif a == RAISE_2POT:
            self.bet(det + pot)
        elif a == ALL_IN:
            self.bet(INIT_CHIPS)

        self.player = (self.player + 1) % NUM_PLAYER
        while self.folds[self.player]:
            self.player = (self.player + 1) % NUM_PLAYER

    def change_state(self):
        if self.num == 1:
            self.pot += np.sum(self.bets)
            self.bets[:] = 0
            return GAME_OVER

        v = self.bets[not self.folds and (self.chips > 0)]
        if np.max(v) != np.min(v):
            return NO_CHANGE

        self.player = self.big_blind
        self.pot += np.sum(self.bets)
        self.bets[:] = 0

        self.step += 1
        return self.step

    def payoff(self):
        max_p = np.max(self.power)
        win = self.power == max_p and not self.folds
        cnt = np.sum(win)
        util = (self.chips - INIT_CHIPS) + (self.pot / cnt) * win
        return util
